### DISCUSSION TITLE
Set HL7 fields in reverse/random order leads to wrong output

### DISCUSSION THREAD

**Discussion by tom08zehn** [nextgenhealthcare/connect#4917] [Q&A] [answered]
I tested in Mirth v3.5.1 and 3.12.0

In the source transformer, I have an outbound message template defined:

```hl7
MSH|^~\&||DICOMPROXY|MIRTH||||ORM^O01||P|2.5
PID||||||||||||||||||||||||||||||||||||||||
PV1||U|||||||||||||||||||||||||||||||||||||
ORC|NW|||||||||||||||||||||||||||||||||||||
OBR||||||||||||||||||||||||||||||||||||||||
ZDS||||||||||||||||||||||||||||||||||||||||
```

In the source transformer script, I set 3 fields in reverse/random order as follows:

```js
tmp['OBR']['OBR.16']['OBR.16.7'] = "Dr.";
tmp['OBR']['OBR.16']['OBR.16.3'] = "John";
tmp['OBR']['OBR.16']['OBR.16.2'] = "Doe";
```

The ER7/XML semantically looks fine (extract of `<OBR.16>` node):

```xml
<OBR.16.7>Dr.</OBR.16.7><OBR.16.3>John</OBR.16.3><OBR.16.2>Doe</OBR.16.2>
```

The rendered `tmp` is wrong and has set:

- OBR.16.7 (correct)
- OBR.16.8 (wrong, instead of OBR.16.3)
- OBR.16.9 (wrong, instead of OBR.16.2)

```hl7
OBR||||||||||||||||^^^^^^Dr.^John^Doe
```

**Why is that and is this intended?**

Please find sample channel attached (created in Mirth v3.12.0).
[HL7 set in random order.zip](https://github.com/nextgenhealthcare/connect/files/7719233/HL7.set.in.random.order.zip)

**Comment by tonygermano** [comment, marked as answer]
It has to do with the way that the mirth serializer works going from xml to hl7. It expects them to appear in the correct order in the xml. If you unable to ensure that elements will be added in the correct order, you can use this code template to fix them https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Fix%20HL7%20Node%20Order

To elaborate a little bit more on why they are getting created out of order, from your template `tmp['OBR']['OBR.16']` will exist, but it will have no children. When you assign to `tmp['OBR']['OBR.16']['OBR.16.7']`, since it doesn't already exist, it will be appended to the `tmp['OBR']['OBR.16']` child list (as the first and only child.) Then when you assign to the also non-existent `tmp['OBR']['OBR.16']['OBR.16.3']`, it gets appended *after* `tmp['OBR']['OBR.16']['OBR.16.7']`.

If your template had at least 6 `^` in OBR.16, then the components would also get pre-created in the correct order and it wouldn't matter which order you assigned to them.

All of the examples below do not depend on this as they would replace any components created by the template if they existed.

**Reply by tonygermano** [reply]
```js
tmp['OBR']['OBR.16']['OBR.16.7'] = "Dr.";
tmp['OBR']['OBR.16']['OBR.16.3'] = "John";
tmp['OBR']['OBR.16']['OBR.16.2'] = "Doe";
```
The above code can also be rewritten as this
```js
tmp['OBR']['OBR.16'] = <OBR.16>
    <OBR.16.2>Doe</OBR.16.2>
    <OBR.16.3>John</OBR.16.3>
    <OBR.16.7>Dr.</OBR.16.7>
</OBR.16>;
```
or assuming your values are not constant and are coming from somewhere else
```js
var doc = {
    firstName: 'John',
    lastName: 'Doe',
    degree: 'Dr.'
};

tmp['OBR']['OBR.16'] = <OBR.16>
    <OBR.16.2>{doc.lastName}</OBR.16.2>
    <OBR.16.3>{doc.firstName}</OBR.16.3>
    <OBR.16.7>{doc.degree}</OBR.16.7>
</OBR.16>;
```

**Reply by tonygermano** [reply]
You could even generalize it like this
```js
function createProvider(fieldName, provider) {
    return <{fieldName}>
        <{fieldName+'.2'}>{provider.lastName}</{fieldName+'.2'}>
        <{fieldName+'.3'}>{provider.firstName}</{fieldName+'.3'}>
        <{fieldName+'.7'}>{provider.degree}</{fieldName+'.7'}>
    </{fieldName}>;
}

tmp['OBR']['OBR.16'] = createProvider('OBR.16', {
    firstName: 'John',
    lastName: 'Doe',
    degree: 'Dr.'
});
```

**Reply by tom08zehn** [reply]
Thanks. I tested `|^^^^^^|` and it worked but wanted to understand if it's intended and why Mirth serializes in such a strange way.

I would expect that Mirth orders the children correctly on its own (without manual correction) before serializing it to HL7.

In my case, the data comes from external and in random order so I cannot control it. The only way is to use the script... many thanks for that!
